src/mapmvt.c

```diff
@@ -416,6 +416,38 @@ static void freeMvtTile(VectorTile__Tile *mvt_tile) {
   msFree(mvt_tile->layers);
 }
 
+/*
+** In case any type of query filter (SLD, OGC, ...) is involved, intercept the
+*msLayerWhichShapes() call.
+*/
+int msMVTWhichShapes(layerObj *layer, rectObj rect, int isQuery) {
+  if (!layer->resultcache)
+    return msLayerWhichShapes(layer, rect, isQuery);
+
+  return MS_SUCCESS;
+}
+
+/*
+** Provides the next shape for the layer. Two possibilities:
+**   - No query filter involved (SLD, OGC, ...), go for msLayerNextShape() since
+*msMVTWhichShapes() let msLayerWhichShapes() pass through.
+**   - A query filter is involved (SLD, OGC, ...), look for resultcache given
+*the last requested index (iShape) -> TODO: may be find a better way.
+**
+*/
+int msMVTGetNextShape(layerObj *layer, shapeObj *shape, int *iShape) {
+  if (!iShape)
+    return MS_FAILURE;
+
+  (*iShape)++;
+
+  if (!layer->resultcache)
+    return msLayerNextShape(layer, shape);
+
+  return msLayerGetShape(layer, shape,
```

Comment:
thre should be some validation that *iShape is in range before accessing layer->resultcache->results[]

Reply:
@rouault : You are 100% correct.
@geographika  : Thank you for taking care